hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -1732,6 +1760,45 @@ int MDMParser::_cbCNUM(int type, const char* buf, int len, char* num)
     return WAIT;
 }
 
+int MDMParser::_cbUCGED(int type, const char* buf, int len, NetStatus* status)
+{
+    if ((type == TYPE_PLUS || type == TYPE_UNKNOWN) && status) {
+        int rsrp;
+        int rsrq_n;
+        unsigned long rsrq_f;
+        // AT+UCGED?
+        // +RSRP: 314,5110,"-102.60",163,5110,"-097.90",173,5110,"-100.40",
+        // +RSRQ: 314,5110,"-19.60",163,5110,"-16.60",173,5110,"-18.60",
+
+        // RSRP maps from dBm [-140,-44] to [0,97]
+        if (sscanf(buf, "\r\n+RSRP: %*d,%*d,\"%d\"",&rsrp) == 1){
+            if (rsrp < -140) {
+                status->rsrp = 0;
+            }
+            else if (rsrp >= -44) {
+                status->rsrp = 97;
+            }
+            else {
+                status->rsrp = rsrp + 140;
+            }
```

Comment:
According to the documentation value of `255` indicates an error, which should be forwarded as-is, as that's also true for CSQ/CESQ. Same for RSRQ.

Reply:
As per this particular AT command, there is no such thing as `255` for `UCGED Mode5`. I tested out removing the antenna and there is only `CME ERROR: no network service` or generic `ERROR`. 
And this is from the manual - 
```
<RSRP> - Reference Signal Received Power (RSRP) as defined in 3GPP TS 36.133 [105] expressed in dBm
<RSRQ> - Reference Signal Received Quality (RSRQ) as defined in 3GPP TS 36.133 [105] expressed in dB
```

But I can default the value to 255, and let the AT command change it when the value is collected from AT command. Please let me know what you think?

Reply:
In any case, when there is no coverage, it will error out and will not give an `OK`...

Reply:
![image](https://user-images.githubusercontent.com/388478/74812661-d0143980-52eb-11ea-8496-981a224aaac2.png)

I was basing my comment on the SARA R4 AT commands reference ^^

Reply:
Let's default to 255 just in case and also handle UCGED potentially returning 255 for us.

Reply:
I defaulted the status->rsrp to 255 before we start calculating the value. And if there are no parsing errors, and if the value reasonably falls in 0 to 97 range which it will for about all int values :D , it will get updated in the `if` loop. 

Please see that: 
-(min_any_value) to -140 dBm : 0
-140 to -45 dBm : 1 to 96
-44 to -(max_any_value) : 97

Reply:
If `255` is returned instead of `-NNN.NN`, I don't think `sscanf(buf, "\r\n+RSRP: %*d,%*d,\"%d.%*u\"",&rsrp) == 1` will be matched and 255 will never be set.  I think you'd potentially have to explicitly scan for it additionally with `sscanf(buf, "\r\n+RSRP: %*d,%*d,\"255\"",&rsrp) == 1` and set it explicitly `status->rsrp = 255;`

I don't think you'd want to implement it as a default case, e.g. `else { status->rsrp = 255; }` because there could be cases where you get back nothing, or garbage and that doesn't exactly equate to 255.

Reply:
@technobly
I do not expect to see a 255 in the command for RSRP/RSRQ because the UCGED command only supposed to give values in dBm, and 255 is an indexed value (not dBm) from CSQ/CESQ command. Do you think the UCGED command will give a 255? 
Either way, it wouldn't hurt to do what you mentioned. I will do this (and handle the case justttt in case 255 doesn't come in quotations) - Does that look about correct? 
```
if ((sscanf(buf, "\r\n+RSRP: %*d,%*d,\"255\"",&rsrp) == 1) || 
                  (sscanf(buf, "\r\n+RSRP: %*d,%*d,255",&rsrp) == 1)) {
       status->rsrp = 255;
}
```

Reply:
There is no need for an additional sscanf for 255, it should be handled correctly by the current one as well. I would also argue that the calling code expects to get 255 if something went wrong (nothing or garbage).